src/mesh/mesh_adaptation/anisotropic_mesh_adaptation.cpp

```diff
@@ -0,0 +1,483 @@
+                flux_hessian_at_istate_idim *= abs(adjoint_gradient[istate][idim]);
+                cellwise_hessian[cell_index] += flux_hessian_at_istate_idim;
+            } //idim
+        } //istate
+    } // cell loop ends
+}
+
+template<int dim, int nstate, typename real, typename MeshType>
+unsigned int AnisotropicMeshAdaptation<dim, nstate, real, MeshType> :: get_iquad_near_cellcenter(
+    const dealii::Quadrature<dim> &volume_quadrature) const
+{
+    dealii::Point<dim,real> ref_center;
+    for(unsigned int idim = 0; idim < dim; ++idim){
+        ref_center[idim] = 0.5;}
+
+    unsigned int iquad_center = 0;
+    real min_distance = 10000.0;
+    for(unsigned int iquad = 0; iquad<volume_quadrature.size(); ++iquad)
+    {
+        const dealii::Point<dim, real> &ref_point = volume_quadrature.point(iquad);
+        const real ref_distance = ref_point.distance(ref_center); 
+        if(min_distance > ref_distance)
+        {
+            min_distance = ref_distance;
+            iquad_center = iquad;
+        }
+    }
+
+    return iquad_center;
+}
+
+// Flux referenced by flux[idof][istate][idim]
+template<int dim, int nstate, typename real, typename MeshType>
+void AnisotropicMeshAdaptation<dim, nstate, real, MeshType> :: get_flux_coeffs(
+    std::vector<std::array<dealii::Tensor<1,dim,real>,nstate>> &flux_at_quads, 
+    const dealii::FEValues<dim,dim> &fe_values_input,
+    const std::vector<dealii::types::global_dof_index> &dof_indices,
+    typename dealii::DoFHandler<dim>::active_cell_iterator cell) const
+{
+    if( ! fe_values_input.get_fe().has_support_points() )
+    {
+        pcout<<"The code here treats flux at support points as flux coeff at idof "
+             <<"which requires an interpolatory FE with support points. Aborting.."<<std::endl;
+        std::abort();
+    }
+        
+    const unsigned int n_dofs_cell = dof_indices.size();
+    dealii::Quadrature<dim> support_pts = fe_values_input.get_fe().get_unit_support_points();
+    dealii::FEValues<dim, dim> fe_values_support_pts(fe_values_input.get_mapping(), fe_values_input.get_fe(),
+                                                     support_pts, dealii::update_values);
+    fe_values_support_pts.reinit(cell);
+    const unsigned int n_quad_pts = fe_values_support_pts.n_quadrature_points;
+    Assert(n_quad_pts == n_dofs_cell, dealii::ExcMessage("n_quad_pts != n_dofs_cell"));
+
+    for(unsigned int iquad = 0; iquad<n_quad_pts; ++iquad)
+    {
+        std::array< real, nstate > soln_at_q;
+        soln_at_q.fill(0.0);
+        for(unsigned int idof = 0; idof < n_dofs_cell; ++idof)
+        {
+            const unsigned int istate = fe_values_support_pts.get_fe().system_to_component_index(idof).first;
+            soln_at_q[istate] += dg->solution(dof_indices[idof]) * fe_values_support_pts.shape_value_component(idof, iquad, istate);
+        }
+        
+        flux_at_quads[iquad] = pde_physics_double->convective_flux(soln_at_q);
```

Comment:
Do we only consider the convective flux here when computing the Hessian of the flux even for cases where diffusion is present?

Reply:
I have implemented INRIA's goal oriented approach for convection dominated flows for now. According to their paper, this approach should work well for Euler's equations with artificial dissipation (and for other cases with negligible diffusion). However, for diffusion dominated flows, their continuous adjoint-based optimization formulation and optimal metric is a bit more involved to implement. Currently, I haven't implemented it but we can extend it to diffusion dominated flows in future.